Introduce `BoxMarker` to improve pretty-printing correctness

nnethercote opened the pull request:
Box opening/closing is really easy to get wrong in the pretty-printers. This PR makes it much harder to get wrong.

r? @Urgau

nnethercote committed:
Introduce `BoxMarker` to pretty-printing.

The pretty-printers open and close "boxes" of text a lot. The open and
close operations must be matched. The matching is currently all implicit
and very easy to get wrong. (#140280 and #140246 are two recent
pretty-printing fixes that both involved unclosed boxes.)

This commit introduces `BoxMarker`, a marker type that represents an
open box. It makes box opening/closing explicit, which makes it much
easier to understand and harder to get wrong.

The commit also removes many comments are on `end` calls saying things
like "end outer head-block", "Close the outer-box". These demonstrate
how confusing the implicit approach was, but aren't necessary any more.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14675450757/job/41190796201) [(plain)](https://github.com/rust-lang/rust/commit/f91c83a8361c6e7746297962128b88f76aa2c482/checks/41190796201/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_ast_pretty v0.0.0 (/checkout/compiler/rustc_ast_pretty)
error: mismatched closing delimiter: `}`
   --> compiler/rustc_ast_pretty/src/pprust/state/expr.rs:456:59
    |
451 |             ast::ExprKind::While(test, blk, opt_label) => {
    |                                                           - closing delimiter possibly meant for this
...
456 |                 self.with_cbox_ibox_print_block_with_attrs(0, 0, blk, attrs, |this| {
    |                                                           ^ unclosed delimiter
...
461 |             }
    |             ^ mismatched closing delimiter

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #140366) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Use `PrintState::head` in `PrintState::block_to_string`.

nnethercote committed:
Inline and remove three pretty-printer methods.

They all have a single call site, aren't that big, and removing them
avoids having to pass some `BoxMarker`s.

nnethercote commented:
Details in the individual commits.

dtolnay commented on compiler/rustc_ast_pretty/src/pprust/state/expr.rs:
```diff
@@ -13,28 +13,28 @@ use rustc_ast::{
-                ast::ExprKind::Block(b, None) => {
-                    self.cbox(0);
-                    self.ibox(0);
+                ast::ExprKind::Block(b, _) => {
```
```suggestion
                ast::ExprKind::Block(b, None) => {
```

It is intentional here that blocks with Some(Label) are not allowed to be used as the else block.

nnethercote replied:
Good catch. I changed `_` to `None` recently in #140280. I suspect this accidental reversion happened while I was resolving conflicts.

dtolnay commented on compiler/rustc_hir_pretty/src/lib.rs:
```diff
@@ -1065,20 +1078,20 @@ impl<'a> State<'a> {
-                hir::ExprKind::Block(b, None) => {
-                    self.cbox(0);
-                    self.ibox(0);
+                hir::ExprKind::Block(b, _) => {
```
```suggestion
                hir::ExprKind::Block(b, None) => {
```

nnethercote commented on compiler/rustc_ast_pretty/src/pp.rs:
```diff
@@ -270,23 +307,30 @@ impl Printer {
         }
+        if cfg!(debug_assertions) && b.ended {
+            panic!("BoxMarker ended more than once");
+        }
```
Mmm, good suggestion.

nnethercote commented:
I addressed @dtolnay's comments.

dtolnay reviewed (approved):
Thank you!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit bb04e11e471b070770edaef785d766255688288d has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay bb04e11e471b070770edaef785d766255688288d -->
<!-- homu: {"type":"Approved","sha":"bb04e11e471b070770edaef785d766255688288d","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
